packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -6267,46 +6266,77 @@ abstract class RenderObjectElement extends Element {
     return ancestor as RenderObjectElement?;
   }
 
-  ParentDataElement<ParentData>? _findAncestorParentDataElement() {
+  List<ParentDataElement<ParentData>> _findAncestorParentDataElements() {
+    // More than one ParentDataWidget can contribute ParentData, but there are
+    // some constraints.
+    // 1. ParentData can only be written by unique ParentDataWidgets.
+    //    For example, two KeepAlive ParentDataWidgets trying to write to the
+    //    same child is not allowed.
+    // 2. The ParentData itself must be compatible with all ParentDataWidgets
+    //    writing to it.
+    //    For example, TwoDimensionalViewportParentData uses the
+    //    KeepAliveParentDataMixin, so it could be compatible with both
+    //    KeepAlive, and another ParentDataWidget with ParentData type
+    //    TwoDimensionalViewportParentData or a subclass thereof.
+    // The first case is verified here. The second is verified in
+    // debugIsValidRenderObject.
     Element? ancestor = _parent;
-    ParentDataElement<ParentData>? result;
+    final List<ParentDataElement<ParentData>> result = <ParentDataElement<ParentData>>[];
+    final Set<Type> ancestorTypes = <Type>{};
     while (ancestor != null && ancestor is! RenderObjectElement) {
       if (ancestor is ParentDataElement<ParentData>) {
-        result = ancestor;
-        break;
+        ancestorTypes.add(ancestor.runtimeType);
```

Comment:
Looks like we only need ancestorTypes to implement a debug check. We should therefore only collect it in debug builds to avoid the extra overhead. In other words, this add should be wrapped in an assert and the variable should probably be called debugAncestorTypes.